lib/AST/Type.cpp

```diff
@@ -3945,49 +3949,71 @@ case TypeKind::Id:
   case TypeKind::TypeAlias: {
     auto alias = cast<TypeAliasType>(base);
     Type oldUnderlyingType = Type(alias->getSinglyDesugaredType());
-    Type newUnderlyingType = oldUnderlyingType.transformRec(fn);
-    if (!newUnderlyingType) return Type();
+    Type newUnderlyingType = oldUnderlyingType.transformRec(fn,
+                                                            forceKeepTypealias);
+    if (!newUnderlyingType)
+      return Type();
+
+    bool anyChange = !oldUnderlyingType->isEqual(newUnderlyingType);
 
     Type oldParentType = alias->getParent();
     Type newParentType;
-    if (oldParentType && !oldParentType->hasTypeParameter() &&
-        !oldParentType->hasArchetype()) {
-      newParentType = oldParentType.transformRec(fn);
-      if (!newParentType) return newUnderlyingType;
+    if (oldParentType) {
+      if (!forceKeepTypealias &&
+          (oldParentType->hasTypeParameter() || oldParentType->hasArchetype()))
+        return newUnderlyingType;
+
+      newParentType = oldParentType.transformRec(fn, forceKeepTypealias);
+      if (!newParentType)
+        return newUnderlyingType;
+      anyChange |= !oldParentType->isEqual(newParentType);
     }
 
+    llvm::SmallVector<Type, 4> replacementTypes;
     auto subMap = alias->getSubstitutionMap();
     if (auto genericSig = subMap.getGenericSignature()) {
+      replacementTypes.reserve(genericSig->getGenericParams().size());
       for (Type gp : genericSig->getGenericParams()) {
```

Comment:
For the "correct" fix, I think you should avoid changing transformRec() altogether. Instead, handle TypeAliasType directly in substType(). To transform the TypeAliasType's SubstitutionMap, call subst() on it, passing in the current substitution and conformance function callbacks.